src/DurableTask.AzureStorage/Partitioning/AppLeaseManager.cs

```
@@ -13,107 +13,369 @@
+                this.starterTokenSource.Dispose();
+            }
+            this.starterTokenSource = new CancellationTokenSource();
+
+            await Task.Run(() => this.AppLeaseManagerStarter(this.starterTokenSource.Token), this.starterTokenSource.Token);
+        }
+
+        async Task AppLeaseManagerStarter(CancellationToken token)
+        {
+            while (!token.IsCancellationRequested)
+            {
+                try
+                {
+                    while (!token.IsCancellationRequested && !await this.TryAquireAppLeaseAsync())
+                    {
+                        await Task.Delay(this.settings.AppLeaseOptions.AcquireInterval);
+                    }
+
+                    await this.StartAppLeaseAsync();
+
+                    await this.AwaitUntilAppLeaseStopped();
+                }
+                catch (Exception e)
+                {
+                    this.settings.Logger.PartitionManagerError(
+                        this.accountName,
+                        this.settings.TaskHubName,
+                        this.workerName,
+                        null,
+                        $"Error in AppLeaseStarter task. Exception: {e}");
+
+                    await Task.Delay(TimeSpan.FromSeconds(10), token);
+                    continue;
+                }
+            }
+        }
+
+        public async Task StopAsync()
+        {
+            if (this.appLeaseIsEnabled)
+            {
+                await this.StopAppLeaseAsync();
+            }
+            else
+            {
+                await this.partitionManager.StopAsync();
+            }
+
+            this.starterTokenSource.Cancel();
+        }
+
+        public async Task ForceChangeAppLeaseAsync()
+        {
+            if (!this.appLeaseIsEnabled)
+            {
+                throw new InvalidOperationException("Cannot force change app lease. UseAppLease is not enabled.");
+            }
+            if (this.isLeaseOwner)
```

Not sure if we should throw an exception here, or just make this a no-op. I could go either way.

---

Right, on second thought I think no-op makes more sense here.